File: 프로그래머스/level3/네트워크.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

vector<int> parent;
vector<int> vi;

void dfs(vector<vector<int>> computers, int p, int c){
    for(int i=0; i<computers[c].size(); i++){
        if(computers[c][i]==1){
            if(vi[i]==1)    continue;
            parent[i] = p;
            vi[i] = 1;
            dfs(computers, p, i);
        }
    }
}
int solution(int n, vector<vector<int>> computers) {
    int answer = 1;
    for(int i=0; i<n; i++){
        parent.push_back(i);
        vi.push_back(0);
    }
    for(int i=0; i<n; i++){
        dfs(computers, i, i);
    }
    sort(parent.begin(), parent.end());
    parent.erase(unique(parent.begin(), parent.end()), parent.end());
    return parent.size();
}

File: 프로그래머스/level3/네트워크.js
function solution(n, computers) {
    let answer = 0;
    let visit = Array(n).fill(0);
    let queue = [];
    
    for(let i=0; i<n; i++){
        if(visit[i]===1)    continue;
        visit[i] = 1;
        answer++;
        for(let j=0; j<n; j++){
            if(computers[i][j]===0||visit[j]===1)    continue;
            visit[j] = 1;
            queue.push(j);
        }
        while(queue.length !== 0){
            let now = queue[0];
            queue.shift();
            for(let j=0; j<n; j++){
                if(computers[now][j]===0||visit[j]===1)  continue;
                visit[j] = 1;
                queue.push(j);
            }
        }
    }
    return answer;
}